pkg/controller/pxc/upgrade.go

```diff
@@ -115,7 +126,156 @@ func (r *ReconcilePerconaXtraDBCluster) updatePod(sfs api.StatefulApp, podSpec *
 	currentSet.Spec.Template.Spec.Affinity = pxc.PodAffinity(podSpec.Affinity, sfs)
 	currentSet.Spec.Template.Spec.Volumes = sfsVolume.Volumes
 
-	return r.client.Update(context.TODO(), currentSet)
+	err = r.client.Update(context.TODO(), currentSet)
+	if err != nil {
+		return fmt.Errorf("update error: %v", err)
+	}
+
+	if !cr.Spec.SmartUpdateEnabled() {
+		return nil
+	}
+
+	err = r.client.Get(context.TODO(), types.NamespacedName{Name: currentSet.Name, Namespace: currentSet.Namespace}, currentSet)
+	if err != nil {
+		return fmt.Errorf("failed to get sate: %v", err)
+	}
+
+	if currentSet.Generation == startGeneration {
+		return nil
+	}
```

Comment:
if smartUpdate exited with error we need to rerun smartUpdate.
```suggestion
	if currentSet.updatedReplicas != nil && currentSet.updatedReplicas >= currentSet.replicas {
		return nil
	}
```